crawlers/mooncrawl/eth_scan.py

```diff
@@ -0,0 +1,184 @@
+        id=id,
+        address=contract_address,
+    )
+    db_session.add(smart_contract)
+    db_session.commit()
+    return id
+
+
+def crawl_step(db_session: Session, contract: VerifiedSmartContract, crawl_url: str):
+    attempt = 0
+    current_interval = 2
+    success = False
+
+    response: Optional[requests.Response] = None
+    while (not success) and attempt < 3:
+        attempt += 1
+        try:
+            response = requests.get(crawl_url)
+            response.raise_for_status()
+            success = True
+        except:
+            current_interval *= 2
+            time.sleep(current_interval)
+
+    if response is None:
+        print(f"Could not process URL: {crawl_url}", file=sys.stderr)
+        return None
+    page = response.json()
+    result = page["result"][0]
+    contract_info = {
+        "data": result,
+        "crawl_version": MOONCRAWL_VERSION,
+        "crawled_at": f"{datetime.now()}",
+    }
+    object_key = f"/v1/etherscan/{contract.address}.json"
+    push_to_bucket(contract_info, object_key)
+
+    eth_address_id = get_address_id(db_session, contract.address)
+
+    eth_label = EthereumLabel(
+        label=ETHERSCAN_SMARTCONTRACTS_LABEL_NAME,
+        address_id=eth_address_id,
+        label_data={
+            "key": object_key,
+            "name": contract.name,
+            "tx_hash": contract.tx_hash,
+        },
+    )
+
+    db_session.add(eth_label)
+    db_session.commit()
+
+
+def crawl(
+    smart_contracts: List[VerifiedSmartContract],
+    interval: float,
+    start_step=0,
+):
+    with yield_db_session_ctx() as db_session:
+        for i in range(start_step, len(smart_contracts)):
+            contract = smart_contracts[i]
+            print(f"Crawling {i+1}/{len(smart_contracts)} : {contract.address}")
+            querry_url = (
+                BASE_API_URL + f"&address={contract.address}&apikey={ETH_SCAN_TOKEN}"
+            )
+            crawl_step(db_session, contract, querry_url)
+            time.sleep(interval)
+        pass
+
+
+def load_smart_contracts() -> List[VerifiedSmartContract]:
+    smart_contracts: List[VerifiedSmartContract] = []
+    s3 = boto3.client("s3")
+    data = s3.get_object(Bucket=bucket, Key="util/verified-contractaddress.csv")
+    for row in csv.DictReader(codecs.getreader("utf-8")(data["Body"])):
+        smart_contracts.append(
+            VerifiedSmartContract(
+                tx_hash=row["Txhash"],
+                address=row["ContractAddress"],
+                name=row["ContractName"],
+            )
+        )
+    return smart_contracts
+
+
+def main():
+    parser = argparse.ArgumentParser(
+        description="Crawls smart contract sources from ethersan.io"
+    )
+    parser.add_argument(
+        "--interval",
+        type=float,
+        default=0.2,
+        help="Number of seconds to wait between requests to the etherscan.io (default: 0.2)",
+    )
+    parser.add_argument(
+        "--offset",
```

Comment:
Good! You are thinking about checkpoints.

Reply:
Might be good to address in the future:

If `crawl_step` fails to crawl a certain smart contract, then we are going to be looking through the logs for the print messages saying that you couldn't crawl a given contract and re-run the crawl only for those contracts.

The current setup where `--offset` is our only checkpoint recovery method will not work well there.

No need to fix in this PR, but please create an issue to improve checkpointing for this crawler.